wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession.swift

```diff
@@ -925,28 +925,45 @@ extension ZMUserSession: ZMSyncStateDelegate {
             context: notificationContext
         ).post()
 
-        let selfClient = ZMUser.selfUser(in: syncContext).selfClient()
-        if selfClient?.hasRegisteredMLSClient == true {
-
-            WaitingGroupTask(context: syncContext) { [self] in
-                // these operations are not dependent and should not be executed in same do/catch
-                do {
-                    // rework implementation of following method - WPB-6053
-                    try await mlsService.performPendingJoins()
-                } catch {
-                    WireLogger.mls.error("Failed to performPendingJoins: \(String(reflecting: error))")
-                }
-                await mlsService.uploadKeyPackagesIfNeeded()
-                await mlsService.updateKeyMaterialForAllStaleGroupsIfNeeded()
+        func performsMLSClientUpdates() async {
+            // these operations are not dependent and should not be executed in same do/catch
+            do {
+                // rework implementation of following method - WPB-6053
+                try await mlsService.performPendingJoins()
+            } catch {
+                WireLogger.mls.error("Failed to performPendingJoins: \(String(reflecting: error))")
             }
-        }
-
-        if mlsFeature.isEnabled {
-            mlsService.commitPendingProposalsIfNeeded()
+            await mlsService.uploadKeyPackagesIfNeeded()
+            await mlsService.updateKeyMaterialForAllStaleGroupsIfNeeded()
         }
 
         WaitingGroupTask(context: syncContext) { [self] in
+            await fetchBackendMLSPublicKeys()
             await fetchAndStoreFeatureConfig()
+
+            let (qualifiedSelfClientID, hasRegisteredMLSClient) = await syncContext.perform {
+                let selfClient = ZMUser.selfUser(in: self.syncContext).selfClient()
+                let hasRegisteredMLSClient = selfClient?.hasRegisteredMLSClient == true
+                return (selfClient?.qualifiedClientID, hasRegisteredMLSClient)
+            }
+
+            if hasRegisteredMLSClient {
+                await performsMLSClientUpdates()
+            } else {
+                // If we discover that
+                // there are MLS public keys on the backend, the MLS feature is enabled and there is no registered MLS client,
+                // we should create one.
+                let needsToRegisterMLSClient = BackendInfo.isMLSEnabled && mlsFeature.isEnabled
```

Comment:
question: I am not sure to fully understand the difference here between BackendInfo.isMLSEnabled && mlsFeature.isEnabled
when isMLSEnabled set ? and why we check both?

Reply:
BackendInfo.isMLSEnabled indicates if the backend supports mls (i.e it has public keys). `misFeature.isEnabled` indicates If mls is enabled for the team. The user can only use mls if it's enabled for both.

Reply:
Yes, just additional info to what John said:

We can create MLS client if it's enabled on the backend side (`GET /mls/public-keys` contains an entry with the key removal) and `mls feature` config is enabled. These 2 conditions must be true for us to create an MLS client.

Reply:
ok thanks make sense, not blocking for now, but what do you think about renaming BackendInfo.isMLSEnabled by BackendInfo.mlsSupported ?